### PR TITLE
Audio asset library testing

### PR DESCRIPTION
**Pull request by pike983**
Adding the audio asset library.

### CODE REVIEW

File: out/build/x64-Debug/thirdparty/glfw3/src/glfw3.pc [hunk 1]
```diff
@@ -0,0 +1,13 @@
+prefix=C:/University/BTech TERM 3/COMP 8551/EngineProjectRedo/GameEngine/out/install/x64-Debug
```
Comment (Snow106) [inline out/build/x64-Debug/thirdparty/glfw3/src/glfw3.pc]: What's the purpose of these files? where these missing from the previous GLFW3 install?

File: Engine/src/Engine/Audio.h [hunk 2]
```diff
@@ -0,0 +1,86 @@
+#pragma once
+#ifndef _AUDIO_H_
+#define _AUDIO_H_
+
+#include <iostream>
+#include "..\irrKlang\irrKlang.h"
+
+// A singleton for controlling sounds throughout the application.
+// Contacts: James
+//
```
Comment (Snow106) [inline Engine/src/Engine/Audio.h]: missing namespace

File: Engine/src/Engine/Audio.h [hunk 3]
```diff
@@ -0,0 +1,86 @@
+#pragma once
+#ifndef _AUDIO_H_
+#define _AUDIO_H_
+
+#include <iostream>
+#include "..\irrKlang\irrKlang.h"
```
Comment (Snow106) [inline Engine/src/Engine/Audio.h]: shouldn't use relative paths to include irrKlang. This should already be available from CMake as an additional include

File: Engine/src/Engine/Audio.h [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+};
+
+
+#pragma region ExampleCode
+
+// For further examples of how to use the ISoundEngine and
+// ISound, view: https://www.ambiera.com/irrklang/docu/index.html
+
+/*
```
Comment (Snow106) [inline Engine/src/Engine/Audio.h]: Coding standards: don't use /* comments.

File: Engine/src/Engine/Audio.cpp [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+// More information on sound effects: https://www.ambiera.com/irrklang/docu/index.html#effects
+// 
+ISound* AudioPlayerSingleton::playSound(const char* soundPath, bool loop, bool startPaused, bool useSoundEffects)
+{
+	ISound* theSound = engine->play2D(soundPath, loop, startPaused, true, ESM_AUTO_DETECT, useSoundEffects);
+	if (theSound != 0)
+	{
+		return theSound;
+	}
+	return NULL;
```
Comment (Snow106) [inline Engine/src/Engine/Audio.cpp]: use nullptr

File: Engine/src/Engine/Audio.cpp [hunk 6]
```diff
@@ -0,0 +1,69 @@
+#include "Audio.h"
+
+using namespace std;
+using namespace irrklang;
+
+AudioPlayerSingleton* AudioPlayerSingleton::instance;
+
+AudioPlayerSingleton* AudioPlayerSingleton::getInstance()
+{
+	if (instance == 0)
```
Comment (Snow106) [inline Engine/src/Engine/Audio.cpp]: Should check for nullptr instead of 0

File: Engine/src/Engine/Audio.cpp [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+// Plays a sound, takes a path to the sound as a const char, a bool if the sound should loop,
+// a bool if the sound should be initialized paused or play on initialization, and a bool for if
+// sound effects should be used. Sound effects need to be added after using this function using
+// the returned ISound*.
+// More information on sound effects: https://www.ambiera.com/irrklang/docu/index.html#effects
+// 
+ISound* AudioPlayerSingleton::playSound(const char* soundPath, bool loop, bool startPaused, bool useSoundEffects)
+{
+	ISound* theSound = engine->play2D(soundPath, loop, startPaused, true, ESM_AUTO_DETECT, useSoundEffects);
+	if (theSound != 0)
```
Comment (Snow106) [inline Engine/src/Engine/Audio.cpp:28]: does the library actually return a 0 or a nullptr if there's an error? If this is to keep parity with the library then it's probably fine

File: Engine/src/Engine/Audio.cpp [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+	{
+		return theSound;
+	}
+	return NULL;
+	//TODO: implement alternate media folder search if multiple media folders
```
Comment (Snow106) [inline Engine/src/Engine/Audio.cpp:33]: It's probably better to leave the TODO here and create a task in Trello with the relevant code to be added

File: Engine/src/Engine/Audio.cpp [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+// Plays a sound in 3D space, takes a const char* path to the sound, 
+// an irrklang::vec3df position (irrklang::vec3df positionName(x,y,z)),
+// a bool if the sound should loop, a bool if the sound should initialize paused,
+// and a bool if the sound should use sound effects. Sound effects need to be added after
+// using this funtion using the returned ISound*.
+// More information on sound effects: https://www.ambiera.com/irrklang/docu/index.html#effects
+// 
+ISound* AudioPlayerSingleton::play3DSound(const char* soundPath, vec3df sound3DPosition, bool loop, bool beginPaused, bool useSoundEffects)
```
Comment (Snow106) [inline Engine/src/Engine/Audio.cpp:44]: is pass by value ok here? (for the vec3def)

File: Engine/src/Engine/Audio.h [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+#include <iostream>
+#include "..\irrKlang\irrKlang.h"
+
+// A singleton for controlling sounds throughout the application.
+// Contacts: James
+//
+class AudioPlayerSingleton
+{
+private:
+	static AudioPlayerSingleton* instance;
```
Comment (Snow106) [inline Engine/src/Engine/Audio.h]: guidelines: m_instance

**Review by Snow106 — CHANGES_REQUESTED** [review]
Everything in out/build/x64-Debug should not be merged into develop.
add relevant filepaths to .gitignore, though I would note that it seems a lot of these were added because certain configurations were changed from VS. Configs in this project should be done through Cmake, reach out if you need help with that

File: Engine/src/Engine/Audio.cpp (hunk 9 above)
Comment (pike983) [reply Engine/src/Engine/Audio.cpp:44]: Yeah, I think it should be. The returned ISound lets the position be changed later.

File: Engine/src/Engine/Audio.cpp (hunk 7 above)
Comment (pike983) [reply Engine/src/Engine/Audio.cpp:28]: The library returns a 0 if it can't play the sound.

File: Engine/src/Engine/Audio.h [hunk 11]
```diff
@@ -4,80 +4,83 @@
...
 #include "irrKlang.h"
 
-// A singleton for controlling sounds throughout the application.
-// Contacts: James
-//
-class AudioPlayerSingleton
-{
-private:
-	static AudioPlayerSingleton* instance;
+namespace IAudio {
```
Comment (Snow106) [inline Engine/src/Engine/Audio.h]: Bracket in new line

Comment (Snow106) [reply Engine/src/Engine/Audio.h]: also namespace should still be Engine

**Review by Snow106 — CHANGES_REQUESTED** [review]

**Review by Snow106 — APPROVED** [review]

**Review by mikesinclair66 — DISMISSED** [review]
src/Engine/Audio.cpp can't use "using namespace example_name";
guidelines

**Comment by Snow106** [comment]
> src/Engine/Audio.cpp can't use "using namespace example_name"; guidelines

While I personally perfer leaving "using" out of .cpp files, the guidelines specifically allow the use of using in implementation files "Writing using namespace in an implementation file is fine though."

**Review by mikesinclair66 — APPROVED** [review]

**Pull request merged**